### PR TITLE
[tests/models] refactor with BoringModel

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Part of https://github.com/PyTorchLightning/pytorch-lightning/issues/4501

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by rohitgr7** [comment]
Need help with the 3 failing GPU tests.

**Comment by rohitgr7** [comment]
Need help with the 3 failing GPU tests.

File: tests/base/develop_pipelines.py
Comment (awaelchli) [inline tests/base/develop_pipelines.py:102]: problem is most likely here.
Perhaps it would be best to pass in `trainer.get_model()` to this function which will correctly unwrap the LightningModule.

**Comment by gianscarpe** [comment]
> Need help with the 3 failing GPU tests.

I can check and help you on that, as I worked on `test_gpu` and developed some of the structure

**Comment by rohitgr7** [comment]
> > Need help with the 3 failing GPU tests.
> 
> I can check and help you on that, as I worked on `test_gpu` and developed some of the structure

that would be great. Thanks @gianscarpe

File: tests/base/develop_pipelines.py
Comment (rohitgr7) [reply tests/base/develop_pipelines.py:102]: only one dp test is failing. `ref_model` is used only to check whether to go for `BoringModel` check function or `EvalModelTemplate`(the old one) check function. Also for dp `trainer.get_model() just returns `trained_model.module` only. There is some other issue here. Need to check.

File: tests/models/test_restore.py
Comment (gianscarpe) [inline tests/models/test_restore.py:441]: Hi @rohitgr7, I'm looking at the errors in , but they're not directly related to GPUs. In particular, error ` FAILED tests/models/test_restore.py::test_dp_resume - TypeError: validation_step() missing 1 required positional argument: ` is related to missing `batch_idx` of the third validation batch when `new_trainer.fit(model)` is called:

```
args = (tensor([[ 0.3957,  0.3204,  1.2297,  1.7594,  2.4312,  1.0279, -0.5972, -0.2981,
          0.7191,  1.0529,  0.4432,  0.3232, -0.2949,  1.3755, -1.6502, -0.3943,
          0.5118, -0.1864,  1.4559,  0.7368,  0.1313, -0.2334, -0.7938, -0.5922,
          0.3301, -0.1205,  0.1629, -0.5484, -0.5139, -0.0191,  0.7912,  0.5872]],
       device='cuda:0'),)
```
First two batches have the batch_idx field. This is the second (idx 1):
```
args = (tensor([[ 0.8496, -0.5701,  1.2025,  1.9781,  0.3139,  0.3777,  0.6508,  0.4263,
          0.6376,  0.2734,  1.3930,  0.7251, -0.4903,  1.3228,  0.1101,  0.7859,
         -0.7761,  0.2667, -0.2144, -0.4566, -0.0672, -0.7441,  0.3575, -0.8824,
         -0.9212, -1.8361, -0.5334, -0.1589, -0.0142, -0.2120, -0.4227, -0.7254]],
       device='cuda:0'), 1)
```
I'm investigating, but it seems strange to me

Comment (gianscarpe) [reply tests/models/test_restore.py:441]: There's another issue, related to parallel gpus. 
When you call 
```
output = trained_model(batch)
```
if `trained_model` is of class `LightningDataParallel` it returns an output like:
```
{'x': tensor([1211.6818], device='cuda:0')}
```
This is nonsense, since the model should return two values when called `forward`, instead of a `dict`.
Do you have any idea?

File: tests/base/develop_pipelines.py
Comment (rohitgr7) [reply tests/base/develop_pipelines.py:102]: It's `BoringModel` for ref_model and `LightiningDataParallel` for trained_model.

File: tests/models/test_restore.py
Comment (rohitgr7) [reply tests/models/test_restore.py:441]: not sure about this. Still not very much comfortable with Lightning distributed code since I don't do distributed training often.

File: .drone.yml
Comment (rohitgr7) [inline .drone.yml:42]:
```suggestion
    - python -m coverage run --source pytorch_lightning -m pytest pytorch_lightning tests/models/test_restore.py -v --durations=25 # --flake8
```

File: tests/base/develop_pipelines.py
Comment (rohitgr7) [inline tests/base/develop_pipelines.py:147]:
```suggestion
        output = trained_model(batch, 0)
        acc = trained_model.module.loss(batch, output)
    else:
        output = trained_model(batch)
        acc = trained_model.loss(batch, output)

```

File: tests/base/develop_pipelines.py
Comment (rohitgr7) [inline tests/base/develop_pipelines.py:144]:
```suggestion
        acc = trained_model(batch, 0)['loss']
```

File: .drone.yml
Comment (rohitgr7) [inline .drone.yml:42]:
```suggestion
    - python -m coverage run --source pytorch_lightning -m pytest pytorch_lightning tests -v --durations=25 # --flake8
```

File: tests/models/test_restore.py
Comment (rohitgr7) [reply tests/models/test_restore.py]: ok.. dp test resolved. 2 to go.

**Comment by gianscarpe** [comment]
Hi @rohitgr7, great work with the refactoring of `BoringModel`. However, I believe it should be better to submit a smaller  PR with less changes to help reviewers. @Borda  what do you think?

**Comment by rohitgr7** [comment]
@gianscarpe most of the changes here are independent so I guess the same amount of work will be required even if I split it into 2 or more. But yeah if others say to split this, it won't be a problem for me :)

Also, did you find anything regarding the remaining 2 failing GPU tests?

**Comment by Borda** [comment]
> Hi @rohitgr7, great work with the refactoring of `BoringModel`. However, I believe it should be better to submit a smaller PR with less changes to help reviewers. @Borda what do you think?

well yes, smaller PR would be easier to read, but since it changes just a few lines it shall be fine after all tests pass :]

**Review by Borda — APPROVED** [review]

**Comment by rohitgr7** [comment]
@Borda Still need some help fixing the 2 remaining GPU tests. It's a bit difficult for me to fix them since I don't have a GPU locally.

**Comment by Borda** [comment]
@gianscarpe @akihironitta mind have look? :]

File: tests/base/boring_model.py
Comment (awaelchli) [inline tests/base/boring_model.py:96]: this is necessary to make  boring model work with amp, because the forward gets wrapped

File: pytorch_lightning/accelerators/dp_accelerator.py
Comment (awaelchli) [inline pytorch_lightning/accelerators/dp_accelerator.py:158]: this is cherry picked from master
#5519

File: pytorch_lightning/trainer/optimizers.py
Comment (awaelchli) [inline pytorch_lightning/trainer/optimizers.py:146]: changes in this file cherry picked from master
#5519

File: tests/base/develop_pipelines.py
Comment (awaelchli) [inline tests/base/develop_pipelines.py:130]: this is an important todo and should be tackled asap after this PR.

File: tests/models/test_restore.py
Comment (awaelchli) [inline tests/models/test_restore.py:438]: @rohitgr7 this was the actual fix for the dp stuff, I reverted the other dp specific changes in the run_prediction function. 
the run prediction function does not need the dp wrapped model.

**Review by awaelchli — COMMENTED** [review]
I still have one test to fix

File: tests/base/develop_pipelines.py
Comment (awaelchli) [reply tests/base/develop_pipelines.py:130]: we need some sort of BoringPredictionModel 
and put `assert isinstance(trained_model, BoringPredictionModel)` into the run_prediction function

Comment (rohitgr7) [reply tests/base/develop_pipelines.py:130]: yes agreed. I think we should have a dataset here that is small but not just completely random. maybe we can have a look at this https://scikit-learn.org/stable/modules/generated/sklearn.datasets.make_classification.html

File: tests/models/test_restore.py
Comment (rohitgr7) [reply tests/models/test_restore.py:438]: I just updated it according to the EvalModel before. but yeah this works too.

File: tests/models/test_amp.py
Comment (awaelchli) [inline tests/models/test_amp.py:194]: @rohitgr7 I had to keep this EvalModelTemplate
To the best of my abilities, I am  just not able to figure out why boring model with multi optimizers does not but eval template does (in the case of apex). 

I propose to leave it for now. In the meantime, I have opened an issue here: #5642

**Comment by awaelchli** [comment]
One more minor annoyance I encountered was the flawed test for the fine tuning callback. I included the fix here, but made a separate PR #5643 too. 

Should be be good now. All gpu tests pass locally on my machine. 
For future PRs  of this kind, I suggest you keep it smaller and do the easy BoringModels first. And tests that have far-reaching implications (like the ones here that involve the helper function for predicting) should be done in separate PRs.

File: tests/models/test_amp.py
Comment (rohitgr7) [reply tests/models/test_amp.py:196]: interesting... on master it's passing with BoringModel. The fix Thomas is working is to ensure that requires_grad for layers remain same.

Comment (awaelchli) [reply tests/models/test_amp.py:196]: ok, maybe the PR with master commits will be merged soon into 1.2 so we will check again when that happens :)

**Comment by Borda** [comment]
just for the record, blocked by #5583

File: tests/models/test_cpu.py
Comment (akihironitta) [inline tests/models/test_cpu.py:203]: noob question: why do we `self.log` in this subclass here instead of in the superclass `BoringModel`? `self.log` seems quite general to me. Is it for efficiency?

Comment (rohitgr7) [reply tests/models/test_cpu.py:203]: yeah maybe since log is not required everywhere and adding that in BoringModel might create an overhead.

Comment (akihironitta) [reply tests/models/test_cpu.py:203]: That makes sense. Thanks :)

File: tests/models/test_restore.py
Comment (gianscarpe) [reply tests/models/test_restore.py]: Did you solve this in the end?

Comment (rohitgr7) [reply tests/models/test_restore.py]: @gianscarpe only 1 test is failing now. As of now it's working with EvalModel but not with BoringModel. `test_amp_with_apex`

Comment (awaelchli) [reply tests/models/test_restore.py]: we fixed it on master btw :)

Comment (gianscarpe) [reply tests/models/test_restore.py]: Perfect! :)

**Review by gianscarpe — APPROVED** [review]

**Comment by carmocca** [comment]
@rohitgr7 This is no longer blocked. Want to update it?

File: tests/models/test_cpu.py
Comment (s-rog) [inline tests/models/test_cpu.py:103]: Do we need `on_epoch` here to make it consistent with previous behavior?

**Review by s-rog — APPROVED** [review]

File: tests/base/develop_pipelines.py
Comment (carmocca) [inline tests/base/develop_pipelines.py:102]: Why do we keep this one?

File: tests/models/test_hparams.py
Comment (carmocca) [inline tests/models/test_hparams.py:532]: Why do we need this class instead of just BoringModel?

File: tests/models/test_cpu.py
Comment (rohitgr7) [reply tests/models/test_cpu.py:103]: for `validation_step`, `on_step=False` and `on_epoch=True` by default.

File: tests/base/develop_pipelines.py
Comment (rohitgr7) [reply tests/base/develop_pipelines.py:102]: still used in other tests. we can remove all eval model related things once all the tests are updated with BoringModel.

File: tests/models/test_hparams.py
Comment (rohitgr7) [reply tests/models/test_hparams.py:532]: true. no need for this. will update.

Comment (rohitgr7) [reply tests/models/test_hparams.py:532]: I guess it's required to check the args for a class that is inherited from a model that requires args.

**Comment by rohitgr7** [comment]
regarding the formatting failures, yapf isn't following pep8.

**Comment by Borda** [comment]
> regarding the formatting failures, yapf isn't following pep8.

fixed

File: tests/models/test_hparams.py
Comment (carmocca) [reply tests/models/test_hparams.py:532]: at least, we don't need to re-define forward, loss, training and test step, ...

File: tests/models/test_amp.py
Comment (kaushikb11) [inline tests/models/test_amp.py:206]: @rohitgr7 Why is this required?

File: tests/models/test_amp.py
Comment (Borda) [inline tests/models/test_amp.py:210]:
```suggestion
        max_steps=5,
```

File: tests/models/test_grad_norm.py
Comment (Borda) [inline tests/models/test_grad_norm.py:39]:
```suggestion
        output = self(batch)
```

File: tests/models/test_hooks.py
Comment (Borda) [inline tests/models/test_hooks.py:126]:
```suggestion
```

File: tests/models/test_hparams.py
Comment (Borda) [inline tests/models/test_hparams.py:561]:
```suggestion
```

File: tests/models/test_restore.py
Comment (Borda) [inline tests/models/test_restore.py:231]: we have now `ClassificationModel`

**Comment by Borda** [comment]
@rohitgr7 I have fixed some todos by using the Classification model which shall converge... mind fix the remaining failing tests? 🐰

File: tests/models/test_hparams.py
Comment (carmocca) [reply tests/models/test_hparams.py:532]: @rohitgr7

File: tests/models/test_amp.py
Comment (rohitgr7) [reply tests/models/test_amp.py]: it's multioptimizer, so by default `training_epoch_end` outputs will be passed differently. so I made it None, since it's not required.

File: tests/models/test_restore.py
Comment (kaushikb11) [inline tests/models/test_restore.py]: @rohitgr7 Why did you remove this?

Comment (rohitgr7) [reply tests/models/test_restore.py]: it's not required here. all we need to ensure that weights are loaded correctly. else I had to update all the generic models added here.

File: tests/models/test_hparams.py
Comment (rohitgr7) [reply tests/models/test_hparams.py:531]: done.

**Review by kaushikb11 — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

File: tests/models/test_restore.py
Comment (rohitgr7) [inline tests/models/test_restore.py:286]:
```suggestion
        tpipes.run_prediction(pretrained_model, dataloader, min_acc=0.1)
```

**Comment by Borda** [comment]
@kaushikb11 mind check the two failing GPU tests?

**Comment by Borda** [comment]
@rohitgr7 @kaushikb11 mind check the last two failing tests?  :]

**Comment by awaelchli** [comment]
Why did you merge it, the tests were not passing

**Comment by Borda** [comment]
ok, that was an issue with switching branches, for while protection was off... :/

**Comment by Borda** [comment]
> Why did you merge it, the tests were not passing

ok, that was an issue with switching branches, for while protection was off... :/

**Comment by awaelchli** [comment]
@Borda what do you suggest? can we undo the commit and then resubmit the PR?

File: tests/core/test_datamodules.py
Comment (SeanNaren) [reply tests/core/test_datamodules.py:394]: I think this is fine, this is on step end not on epoch end, so the underlying metric will just accumulate via the `update` method. Only at epoch end will we call `compute` I think

This is to get around the issue that training step is within the DP scatter op, where the outputs will be on multiple devices

**Comment by Borda** [comment]
@kaushikb11 is already working on that fix...

File: tests/core/test_datamodules.py
Comment (awaelchli) [reply tests/core/test_datamodules.py:394]: Yes I didn't see it's using a metric. 
My concern was about the return None value here. But if that is handled internally then I guess no issue

**Comment by awaelchli** [comment]
Me and rohit were also working on the fix before it was merged

**Pull request merged**
